pg_search/src/postgres/storage/fsm.rs

```diff
@@ -34,413 +35,1263 @@ enum FSMBlockKind {
+    }
+}
+
+/// The [`v2`] FreeSpaceManager is a fixed-size AVL tree laid out as an array on the FSM's first block.
+///
+/// Each entry in the AVL tree is called a [`Slot`] and each slot has a key, value, and tag.  The key
+/// is a 64bit [`pg_sys::FullTransactionId`], the value is actually the Rust unit type ([`()`]), and
+/// the tag is a [`pg_sys::BlockNumber`].
+///
+/// Each slot gets its own statically assigned tag value when a relation's FSM is first initialized.
+/// This value is the block number that represents the start of the list of block numbers that are
+/// free for the xid key stored in that slot.  This value stays with the slot, even as the tree is
+/// mutated as generally speaking, with an array-backed AVL tree entries don't move slots, only their
+/// left/right pointers are updated during rebalancing.  Special consideration is made for deleting
+/// an entry, however, when the key/value is moved to a new slot -- in this case the tag values are
+/// swapped until the entry is finished moving.  Essentially, tags have an affinity for their key/value
+/// as long as the slot is occupied.
+///
+/// The FSM has two user-facing operations: extend and drain.
+///
+/// # Extend
+///
+/// The caller provides both a transaction id and an iterator of free blocks to associate with that
+/// transaction.  That transaction indicates the point in time in which those blocks are usable.  Any
+/// future transaction `>=` to that transaction will be able to use those blocks.
+///
+/// During an extension a share lock is acquired on the FSM's root page (the AVL tree) and the
+/// specified transaction id is searched.  If it's found, then an exclusive lock is taken on the block
+/// represented by that xid's `tag` -- the blocklist -- and the blocklist is extended from head-to-tail,
+/// filling in gaps along the way.  However, if a full page is encountered, then a shortcut is taken
+/// and a new blocklist is linked into the existing list at that point.
+///
+/// Concurrent extensions are allowed on the same xid, blocking one block in the block list at a time.
+/// However, in practice almost every call to [`FreeSpaceManager::extend_with_when_recyclable`] uses
+/// the current transaction id, and since each transaction has its own unique id, there'll be no blocking.
+///
+/// Sometimes a caller will use the result of [`pg_sys::ReadNextFullTransactionId()`] instead of the
+/// current transaction id.  This can cause concurrent extensions of the same transaction id.  This
+/// is fine and generally a rare situation.
+///
+/// If the provided transaction id is not found, then the shared lock on the FSM root is upgraded to
+/// an exclusive lock and the transaction id is inserted.  It is possible that by this time the
+/// transaction id now exists.  In either case, the `tag` value for the new (or now-existing) transaction
+/// id's slot is used as the blocklist, and it is extended.
```

Comment:
Would the effect of this be that if a transaction doesn't free very many blocks, it might have a very small blocklist/freelist? For example, a transaction which does a merge for 10, 1k segments would result in a freelist of length 80.

Would it be worth rounding down in some cases, so that freelists are less likely to be very small?

Reply:
It wouldn't be rounding "down", but rounding "up" to a newer transaction xid and that has the downside of shifting the timeline of when those blocks can be reused.

I don't think in practice this is worth worrying about unless future profiling highlighted this case.